#1: Allow shortcuts to work inside input fields with flag

zshall opened the pull request:
* Disable the default filter that Keymaster uses, processing key commands on all keypresses.
* Changing the signature of `@combo()` to declare a list of shortcuts as a single parameter.
* Adding a second parameter, `runInsideInputs`, a flag to determine whether we should filter or not.
* Adding an `if` block inside the Keymaster handler to `return true` if we should be filtering inside inputs, and the default filter detects that we are currently in an input.

I added examples of both to the `dev-app`.

zshall committed:
#1 --- set up dev app for testing on input fields

zshall committed:
#1 --- disable Keymaster's built-in filter

zshall committed:
#1 --- enable Chrome debugging in Visual Studio Code

zshall committed:
#1 --- make shortcuts a single argument, so we can add more args later

zshall committed:
#1 --- create a flag, runInsideInputs, which will disable default keymaster filtering

zshall committed:
#1 --- fixed linting errors, unit tests updated w/ single parameter

zshall commented:
I'm not familiar with the testing system used here and am not yet sure how to update the unit tests so that we can simulate focusing an input box. Looking into it now.

zshall commented:
I'm not sure how to proceed in testing this. Here's what I've tried:

I've added an HTML input `<input id="test-input" />` to line 44 in the spec file:

`@inlineView('<template><input id="test-input" /></template>')`

I've also added a test that is supposed to focus this test input and sends keys:

~~~js
      nq(() => {
        const testInput = document.getElementById('test-input');
        testInput.focus();
        // ctrl+f
        keydown(KEYS.ctrl); keydown(70, [KEYS.ctrl]); keyup(70); keyup(KEYS.ctrl);

        // no more trigger after focusing input
        expect(results).toEqual(['findIt', 'findIt', 'copyIt']);
      });
~~~

This doesn't seem to work, as the event is not being filtered out.

3cp commented on src/index.js:
```diff
@@ -58,15 +69,19 @@ function eachMethod(obj, callback) {
 }
 
-export function combo(...shortcuts) {
+export function combo(shortcuts, runInsideInputs) {
```
This method signature change is a breaking change. Suggest to keep old signature, then check last param.

```js
combo(...shortcuts) {
  let runInsideInputs = false;
  if (shortcuts.length && typeof shortcuts[shortcuts.length - 1] === 'boolean') {
    runInsideInputs = shortcuts.pop();
  }
  //...
}
```

That will enable
```js
@combo('ctrl+f', 'command+f', true)
```

3cp commented on src/index.js:
```diff
@@ -11,14 +11,20 @@ export function configure() {
   const attached = Controller.prototype.attached;
   const detached = Controller.prototype.detached;
 
+  key.filter = () => true; // disable Keymaster's built-in filter
```
Remove duplicated defaultKeymasterFilter, just get it from keymaster.

```js
const defaultKeymasterFilter = key.filter;
key.filter = () => true;
```

3cp replied:
Since we have local copy of keymaster, the other way we can clean it up is to totally remove filter from keymaster file, then use your code here.

3cp commented on src/index.js:
```diff
@@ -11,14 +11,20 @@ export function configure() {
-        key(value.combo, e => {
+        key(value.combo.shortcuts, e => {
+          if (!value.combo.runInsideInputs && !defaultKeymasterFilter(e)) {
+            return true;
```
This can be just `return;` I think.
The return value here is not used by aurelia.

3cp commented:
Thx, I didn't realise is keymaster blocking the events.
For the tests, try a delay after focus.
```js
ng(() => do_focus);
ng(() => do_keypress...);
// following existing test code, the "expect" should be delayed too
ng(() => expect...);
```

3cp commented:
If tests cannot work, just push them up. I will review in the long weekend.

zshall committed:
#1 --- variable arguments, last one is runInsideInputs flag if boolean

zshall committed:
#1 --- use Keymaster default filter from Keymaster library

zshall committed:
#1 --- Change `return true` to `return` for Keymaster handler function

zshall committed:
#1 --- make unit tests pass with input focused

When we dispatch the key event, use `document.activeElement` as the event target.

zshall committed:
#1 --- keyUp should use activeElement too

zshall commented:
@3cp addressed your comments. Also figured out how to get the tests to work. `document` was the event target whenever the keys events were dispatched; now it's `document.activeElement`, so our `.focus()` calls work properly. Also used `ref` for the test input element so that we don't need to use `getElementById()`

3cp commented:
Thx!

3cp commented:
Released v1.1.0

3cp commented:
I am using it. Love it!
